File: python/src/v3/access_token.py
import requests

from access_token_common import AccessTokenCommon
from user_auth import get_auth_code

class AccessToken(AccessTokenCommon):
    def __init__(self, api_config, name=None):
        super().__init__(api_config, name)

    def oauth(self, scopes=None, refreshable=True):
        """
        Execute the OAuth 2.0 process for obtaining an access token.
        For more information, see IETF RFC 6749: https://tools.ietf.org/html/rfc6749
        and https://developers.pinterest.com/docs/redoc/#tag/Authentication
        """
        print('getting auth_code...')
        auth_code = get_auth_code(self.api_config, scopes=scopes, refreshable=refreshable)
        print(f'exchanging auth_code for {self.name}...')
        self.exchange_auth_code(auth_code)

    def exchange_auth_code(self, auth_code):
        """
        Call the Pinterest API to exchange the auth_code (obtained by
        a redirect from the browser) for the access_token and (if requested)
        refresh_token.
        """
        put_data = {'code': auth_code,
                    'redirect_uri': self.api_config.redirect_uri,
                    'grant_type': 'authorization_code'}
        if (self.api_config.verbosity >= 2):
            print('PUT', self.api_config.api_uri + '/v3/oauth/access_token/')
            if (self.api_config.verbosity >= 3):
                self.api_config.credentials_warning();
                print(put_data)
        response = requests.put(self.api_config.api_uri + '/v3/oauth/access_token/',
                                headers=self.auth_headers, data=put_data)
        unpacked = self.unpack(response)
        print('status: ' + unpacked['status'])

        """
        The scope returned in the response includes all of the scopes that
        have been approved now or in the past by the user.
        """
        print('scope: ' + unpacked['scope'])
        self.access_token = unpacked['access_token']
        self.refresh_token = unpacked['data'].get('refresh_token')
        self.scopes = unpacked['scope']
        if self.refresh_token:
            print('received refresh token')

    def refresh(self):
        if not self.refresh_token:
            raise RuntimeError('AccessToken does not have a refresh token')
        print(f'refreshing {self.name}...')
        put_data = {'grant_type': 'refresh_token',
                    'refresh_token': self.refresh_token}
        if (self.api_config.verbosity >= 2):
            print('PUT', self.api_config.api_uri + '/v3/oauth/access_token/')
            if (self.api_config.verbosity >= 3):
                self.api_config.credentials_warning();
                print(put_data)
        response = requests.put(self.api_config.api_uri + '/v3/oauth/access_token/',

                                headers=self.auth_headers, data=put_data)
        unpacked = self.unpack(response)
        self.access_token = unpacked['access_token']

File: nodejs/src/v3/access_token.js
import got from 'got'

import {AccessTokenCommon} from '../access_token_common.js'
import get_auth_code from '../user_auth.js'

export class AccessToken extends AccessTokenCommon {

  constructor(api_config, {name}) {
    super(api_config, {name: name});
  }

  /**
   * Execute the OAuth 2.0 process for obtaining an access token.
   * For more information, see IETF RFC 6749: https://tools.ietf.org/html/rfc6749
   * and https://developers.pinterest.com/docs/redoc/#tag/Authentication
   *
   * Constructor may not be async, so OAuth must be performed as a separate method.
   */
  async oauth({scopes=null, refreshable=true}) {
    console.log('getting auth_code...');
    const auth_code = await get_auth_code(this.api_config,
                                          {scopes:scopes, refreshable:refreshable});

    console.log('exchanging auth_code for access_token...');
    try {
      const put_data = {
        code: auth_code,
        redirect_uri: this.api_config.redirect_uri,
        grant_type: 'authorization_code'
      };
      if (this.api_config.verbosity >= 2) {
        console.log('PUT', `${this.api_uri}/v3/oauth/access_token/`);
        if (this.api_config.verbosity >= 3) {
          this.api_config.credentials_warning();
          console.log(put_data);
        }
      }
      const response = await got.put(`${this.api_uri}/v3/oauth/access_token/`, {
        headers: this.auth_headers, // use the recommended authorization approach
        json: put_data,
        responseType: 'json'
      })
      this.print_response(response);
      console.log('status:', response.body.status);

      // The scope returned in the response includes all of the scopes that
      // have been approved now or in the past by the user.
      console.log('scope:', response.body.scope);
      this.scopes = response.body.scope;
      this.access_token = response.body.access_token;
      this.refresh_token = response.body.data.refresh_token;
      if (this.refresh_token) {
        console.log('received refresh token');
      }
    } catch (error) {
      this.print_and_throw_error(error);
    }
  }

  async refresh() {
    // There should be a refresh_token, but it is best to check.
    if (!this.refresh_token) {
      throw 'AccessToken does not have a refresh token';
    }

    console.log('refreshing access_token...');
    try {
      const put_data = {
        grant_type: 'refresh_token',
        refresh_token: this.refresh_token
      };
      if (this.api_config.verbosity >= 2) {
        console.log('PUT', `${this.api_uri}/v3/oauth/access_token/`);
        if (this.api_config.verbosity >= 3) {
          this.api_config.credentials_warning();
          console.log(put_data);
        }
      }
      const response = await got.put(`${this.api_uri}/v3/oauth/access_token/`, {
        headers: this.auth_headers,
        json: put_data,
        responseType: 'json'
      })
      this.print_response(response)
      this.access_token = response.body.access_token;
    } catch (error) {
      this.print_and_throw_error(error);
    }
  }
}
